sklearn/metrics/_pairwise_distances_reduction.pyx

```diff
@@ -244,12 +237,457 @@ cdef class PairwiseDistancesReduction:
         -------
         True if the PairwiseDistancesReduction can be used, else False.
         """
-        # TODO: support sparse arrays and 32 bits
+        dtypes_validity = X.dtype == Y.dtype and Y.dtype == np.float64
         return (get_config().get("enable_cython_pairwise_dist", True) and
-                not issparse(X) and X.dtype == np.float64 and
-                not issparse(Y) and Y.dtype == np.float64 and
+                not issparse(X) and not issparse(Y) and dtypes_validity and
                 metric in cls.valid_metrics())
 
+
+class PairwiseDistancesArgKmin(PairwiseDistancesReduction):
+    """Compute the argkmin of row vectors of X on the ones of Y.
+
+    For each row vector of X, computes the indices of k first the rows
+    vectors of Y with the smallest distances.
+
+    PairwiseDistancesArgKmin is typically used to perform
+    bruteforce k-nearest neighbors queries.
+
+    This class is not meant to be instanciated, one should only use
+    its :meth:`compute` classmethod which handles allocation and
+    deallocation consistently.
```

Comment:
[`abstractclassmethod` is deprecated since python 3.3](https://docs.python.org/3/library/abc.html#abc.abstractclassmethod). I will use the combination of `classmethod` and `abstractmethod` as indicated.